psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1012 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Psi4: an open-source quantum chemistry software package
+ *
+ * Copyright (c) 2007-2022 The Psi4 Developers.
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This file is part of Psi4.
+ *
+ * Psi4 is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * Psi4 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "jk.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/electrostatic.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/liboptions/liboptions.h"
+#include "psi4/lib3index/dftensor.h"
+
+#include <vector>
+#include <map>
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+SharedMatrix compute_numeric_overlap(std::shared_ptr<DFTGrid> grid, std::shared_ptr<BasisSet> primary) {
```

Comment:
Why are we bothering with `shared_ptr`? Thanks to return value optimization, there's no performance hit for returning a `Matrix` rather than a pointer to a Matrix, and we never need to do memory management operations on these objects.